src/Symfony/Component/OptionsResolver/OptionsResolver.php

```diff
@@ -1040,24 +1036,31 @@ public function offsetGet($option, bool $triggerDeprecation = true)
             $invalidTypes = [];
 
             foreach ($this->allowedTypes[$option] as $type) {
-                if ($valid = $this->verifyTypes($type, $value, $invalidTypes)) {
+                if ($valid = \is_scalar($type) && $this->verifyTypes($type, $value, $invalidTypes)) {
                     break;
                 }
             }
 
             if (!$valid) {
+                $fmtProvidedTypes = implode('|', array_keys($invalidTypes));
+                if (!$fmtProvidedTypes) {
+                    $invalidTypes = $this->allowedTypes[$option];
+                    array_walk($invalidTypes, static function (&$item) {
+                        $item = \is_scalar($item) ? null : \gettype($item);
+                    });
+                    throw new InvalidOptionsException(sprintf('The option "%s" contains invalid non-scalar type definitions : "%s".', $this->formatOptions([$option]), implode(', ', array_unique($invalidTypes))));
```

Comment:
this looks wrong to me. At this point, what is invalid are not the provided options but the configuration of the OptionsResolver

Reply:
Thank you for your comments.
 - Can you point me which exception should we throw ?
 - At this point we are not validating anything, we are just creating a exception message with all the bad types

Reply:
the code you are adding is validating the configuration of the allowed types instead of validating the value.

`setAllowedTypes` has a signature defined as `@param string|string[]` (and strings are turned into a single-element array containing them). So the code here has no reason to deal with other stuff than strings.

Either we keep the code as is because passing an argument that does not satisfy the type is not covered by any guarantee from Symfony (and this includes not guaranteeing how it will fail btw), and you can rely on static analysis tools like psalm or phpstan (or most IDEs) to detect the invalid usage. The benefit of this option is that the OptionsResolver is often used in a hot path (for big complex forms, we might be using hundreds of OptionsResolver instance getting configured and used to resolve options).
Or we decide to add a runtime validation guarding against invalid values not matching our signature. And then, the only place where this can be done in a meaningful way is in `setAllowedTypes`, not much later.

And the current rule for the Symfony codebase is to go for the first option. We don't add defensive checks against values not matching the signature everywhere in Symfony because this would add a non-negligeable performance impact (all simple cases are covered by native parameter types, so the remaining defensive checks would mostly be on arrays which require loops to check them)

Reply:
I would be in favor of closing, yes

Reply:
Closed.
Thanks for taking time explaining